### ISSUE TITLE
Inserting intersections with 'Function' while narrowing breaks code

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#26970]
@ahejlsberg and I ran into something like this in the RWC. This currently repros on the nightlies.

```ts
function foo(x: Object | (() => string)) {
    if (typeof x === "function") {
        // Previously had no errors
        x();
    }
}
```

**Expected**: No error

**Actual**:

```
[ts] Cannot invoke an expression whose type lacks a call signature. Type '(() => string) | (Object & Function)' has no compatible call signatures.
```

Potentially related to #25243.

**Comment by weswigham** [comment]
Mmmm kinda? Previously we only selected applicable union members, while now we also narrow the not-immediately-applicable ones like we do in the non-union case. In the case of primitives and literals, they evaporate into `never`s and narrow away, while `Object` narrows to `Object & Function` (which would reduce to just `Function` if we simplified non-primitives) which, given our inability to consider function-ish things as actually potentially callable, causes the issue. So it's an older issue but cropping up in a new place due to the change

**Comment by DanielRosenwasser** [comment]
What does not *immediately* applicable mean? Why does this case work?

```ts
function foo(x: { blah: number } | (() => string)) {
    if (typeof x === "function") {
        x()
    }
}
```

**Comment by RyanCavanaugh** [comment]
`Object | (() => string)` is subject to subtype reduction so this example is pretty sketchy to begin with.

**Comment by weswigham** [comment]
So it's probably worth mentioning that `Object & Function` and `() => string` on their own are both callable. The first is resolved as an untyped call, while the second is strong. #27422 is related (which is making the error more clear), and so is #7294, #9239, and #10620 (which would involve removing the error).

**Comment by rafsawicki** [comment]
This caught me off guard today, after upgrading to typescript 3.1. I think it's not uncommon to have a parameter which is either a value or a function returning it - in our company we use it often for helper functions in tests. From a logical point of view it's strange that I can't call something I've just checked to be a function, especially since as @DanielRosenwasser showed it still works in some cases. So for me it's definitely a bug and not just something that requires a better error message.

**Comment by BernieSumption** [comment]
There is a workaround that worked for our use case at least and has allowed us to upgrade to 3.1: use `x instanceof Function` instead of `typeof x === "function"`.

This breaks in 3.1.0

```ts
let f: {} | (() => {}) = null as any;
let s = typeof f === "function" ? f() : f;
//                                ^^^ error
```

This does not

```ts
let f: {} | (() => {}) = null as any;
let s = f instanceof Function ? f() : f;
```
